src/lib/auth/oauth-client.ts

```diff
@@ -0,0 +1,161 @@
+      client_secret: clientSecret,
+      code: this.auth_code,
+      redirect_uri: `${REDIRECT_URI}:${this.port}`,
+      code_verifier: this.code_verifier,
+      ttl: now.toISOString(),
+    };
+    return fetch(`${accountURL}/api/v1/oauth/token`, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/x-www-form-urlencoded",
+      },
+      body: new URLSearchParams(params),
+    });
+  }
+
+  private _generateCSRFToken(): string {
+    return Buffer.from(randomBytes(20)).toString("base64url");
+  }
+
+  private _handleRequest(req: IncomingMessage, res: ServerResponse) {
+    const allowedOrigins = [
+      "http://localhost:3005",
+      "http://127.0.0.1:3005",
+      "http://dashboard.fauna.com",
+      "http://dashboard.fauna-dev.com",
+      "http://dashboard.fauna-preview.com",
+    ];
+    const origin = req.headers.origin || "";
+
+    if (allowedOrigins.includes(origin)) {
+      res.setHeader("Access-Control-Allow-Origin", origin);
+      res.setHeader("Access-Control-Allow-Methods", "GET");
+      res.setHeader("Access-Control-Allow-Headers", "Content-Type");
+    }
+
+    let errorMessage = "";
+
+    if (req.method === "GET") {
+      const parsedUrl = url.parse(req.url || "", true);
+      if (parsedUrl.pathname !== "/") {
+        errorMessage = "Invalid redirect uri";
+        this.closeServer();
+      }
+      const query = parsedUrl.query;
+      if (query.error) {
+        errorMessage = query.error.toString();
+        this.closeServer();
+      }
+      if (query.code) {
+        const authCode = query.code;
+        if (!authCode || typeof authCode !== "string") {
+          errorMessage = "Invalid authorization code received";
+          this.server.close();
+        } else {
+          this.auth_code = authCode;
+          if (query.state !== this.state) {
+            errorMessage = "Invalid state received";
+            this.closeServer();
+          }
+          // TODO: Send them to a nice page that shows auth is complete and they can close the window.
+          res.writeHead(200, { "Content-Type": "text/html" });
+          res.end();
+          this.server.emit("auth_code_received");
+          this.closeServer();
+        }
+      }
+    } else {
+      errorMessage = "Invalid request method";
+      this.closeServer();
+    }
+    if (errorMessage) {
+      console.error("Error during authentication:", errorMessage);
+    }
+  }
+
+  private _getFreePort(): Promise<number> {
+    return new Promise((res, rej) => {
+      const srv = net.createServer();
+      srv.listen(0, () => {
+        const port = srv.address() as AddressInfo;
+        srv.close((err) => {
+          if (err) rej(err);
+          res(port.port);
+        });
+      });
+    });
+  }
+
+  public async start() {
+    try {
+      this.port = await this._getFreePort();
+      this.server.listen(this.port);
+    } catch (e: any) {
+      console.error("Error starting loopback server:", e.message);
+    }
+  }
```

Comment:
this will cause race conditions. the docs say a 'listening' event should be emitted once its ready, so i think something like this should work (that `event == 'listenting'` bit isn't right, but it should be close i think):
```suggestion
  private _listenOnFreePort(): Promise<number> {
    return new Promise((res, rej) => {
      this.server.listen(0, (event) => {
        if (event == 'listening') {
          this.port = (srv.address() as AddressInfo).port;
          res();
        }
      });
    });
  }

  public async start() {
    try {
      await this._listenOnFreePort();
    } catch (e: any) {
      console.error("Error starting loopback server:", e.message);
    }
  }
```

Reply:
@macmv thanks I just committed something similar and it's working